### PR TITLE
[SPARK-24155][ML] Instrumentation improvements for clustering

### PR DESCRIPTION
**Pull request by lu-wang-dl**
## What changes were proposed in this pull request?

changed the instrument for all of the clustering methods

## How was this patch tested?

N/A

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala [hunk 1]
```diff
@@ -423,6 +423,8 @@ class GaussianMixture @Since("2.0.0") (
     val summary = new GaussianMixtureSummary(model.transform(dataset),
       $(predictionCol), $(probabilityCol), $(featuresCol), $(k), logLikelihood)
     model.setSummary(Some(summary))
+    instr.logNamedValue("logLikelihood", logLikelihood)
+    instr.logNamedValue("clusterSizes", summary.clusterSizes.toString)
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: The `clusterSizes.toString` will get an unreadable object address string.
We need to print the content of the array. I suggest add a method like:
`def logNamedArray[T](array: Array[T]): Unit`

File: mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala [hunk 2]
```diff
@@ -378,6 +378,7 @@ class KMeans @Since("1.5.0") (
       model.transform(dataset), $(predictionCol), $(featuresCol), $(k))
 
     model.setSummary(Some(summary))
+    instr.logNamedValue("clusterSizes", summary.clusterSizes.toString)
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/KMeans.scala]: ditto.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala (hunk 1 above)
Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/GaussianMixture.scala]: also ok I think.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala [hunk 3]
```diff
@@ -278,6 +279,7 @@ class BisectingKMeans @Since("2.0.0") (
     val summary = new BisectingKMeansSummary(
       model.transform(dataset), $(predictionCol), $(featuresCol), $(k))
     model.setSummary(Some(summary))
+    instr.logNamedValue("clusterSizes", summary.clusterSizes.mkString(", "))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: This requires parsing the result string if we want to do some analysis. How do we do this in other places? Do we always log in JSON format, e.g.? cc: @WeichenXu123 @jkbradley

Comment (jkbradley) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: We log Params in JSON format, but we don't have a precedent for logging Array values.  I'd be OK with ducking the issue by not logging this value, or with logging it as JSON.  If the latter, then we could add a method Instrumentation.logNamedValue taking Array types.

Comment (mengxr) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: sg. @ludatabricks could you update the value to `mkString("[", ",", "]")` for now and leave a TODO to extend `logNamedValue`?

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

**Pull request closed without merging**